[Console] introduce `__toString` `method` annotation to `InputInterface`

boesing opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4 
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #42042
| License       | MIT
| Doc PR        | none yet, please ping me if this will be necessary

<!--
Replace this notice by a short README for your feature/bugfix. This will help people
understand your PR and can be used as a start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Never break backward compatibility (see https://symfony.com/bc).
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against branch 5.x.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
-->

As a preparation for 6.0, we are adding `__toString` as method declaration. This avoids having BC-breaks in upstream projects as suggested by [@wouterj](https://github.com/symfony/symfony/issues/42042#issuecomment-877409203).
In 6.0, the `Stringable` interface will be implemented instead.

Fixes #42042


If this feature will be accepted, I'm willing to create the PR for 6.0 which adds the `Stringable` interface. 👍🏼

boesing committed:
feature: introduce `__toString` `method` annotation

As a preparation for 6.0, we are adding `__toString` as method declaration. This avoids having BC-breaks in upstream projects as suggested by @wouterj.
In 6.0, the `Stringable` interface will be implemented instead.

Fixes #42042

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add `__toString` to test asset

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

nicolas-grekas commented:
Please add a changelog entry in the component.
It could also be useful to document what the method is supposed to return, because it's not obvious to me at least.

boesing commented:
@nicolas-grekas 🤨 it already exists, I just added it to the interface. could you give me a little hint on how you expect that being documented?
Will create it afterwards.

nicolas-grekas commented:
If we add the method, it means we're expect implementations to return something useful.
We should tell what kind of useful output is expected here.
You could take inspiration from what existing implementations do already I suppose.
I don't know (and I didn't had a look :) )

derrabus commented on src/Symfony/Component/Console/Tests/EventListener/ErrorListenerTest.php:
```diff
@@ -143,4 +143,9 @@ public function getParameterOption($values, $default = false, $onlyParams = fals
     {
     }
+
+    public function __toString()
```
```suggestion
    public function __toString(): string
```

boesing replied:
shall I add types to all other methods which lack them or just to my newly added one?

boesing committed:
qa: add return type to test asset

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
documentation: add `__toString` method description on what to expect when calling that method

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
documentation: add `CHANGELOG` entry for the new `InputInterface::__toString` method annotation

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
I've created a `CHANGELOG` entry as well as the method description.
If something is missing, please let me know.

@derrabus I've added the return type-hint to the newly created method within the test-asset.

boesing commented on src/Symfony/Component/Console/Input/InputInterface.php:
```diff
@@ -18,6 +18,9 @@
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @method string __toString() Returns a stringified representation of the args passed to the command.
+ *                             InputArguments MUST be escaped as well as the InputOption values passed to the command.
```
I could probably link to the existing implementations for further clarifications? Or is this assumed to be enough?

OskarStark replied:
I am fine with it

OskarStark commented:
Could you please open a PR on the `symfony/docs` repo and add a test case? Thanks

boesing commented:
So just to clarify so I understand the open TODOs:

I have added exactly zero new functionality here and there are existing unit tests for those `__toString` methods in current implementations. What kind of unit tests do you want me to create? 

Regarding the docs - I will check that out tomorrow to see how this project is structured. Do you want me to add the description I've added to the interface to the documentation or  do you have something specific in mind?

OskarStark commented on src/Symfony/Component/Console/Input/InputInterface.php:
```diff
@@ -18,6 +18,9 @@
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @method string __toString() Returns a stringified representation of the args passed to the command.
```
AFAIK if someone does not implement this method, a deprecation is thrown.

I would like to have a test case which asserts this deprecation (a missing implementation) and has the "@group legacy".

Not sure I missed sth.

boesing replied:
I'll see how to achieved that. Thanks for clarification. 
~The unit test to verify the deprecation should be part of the docs repo?~

derrabus commented on src/Symfony/Component/Console/CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 ---
 
  * Add `TesterTrait::assertCommandIsSuccessful()` to test command
+ * Add `InputInterface::__toString()` method annotation. This will be replaced by the extension of the `Stringable` interface in 6.0.
```
```suggestion
 * Add method `__toString()` to `InputInterface`
```

derrabus reviewed (changes requested):
Please also adjust `UPGRADE-5.4.md` and `UPGRADE-6.0.md`. If we introduce a deprecation, it must be documented there.

nicolas-grekas commented:
Friendly ping @boesing

boesing commented:
Thanks for pinging me, @nicolas-grekas
I dont think I will work more on this because of the following reason:
1. I dont really need the feature that much that I want to work on multiple repositories to add a method which is already present in almost all implementations
2. I have plenty of open work to do in another framework to be ready for PHP 8.1 and thus the amount of time is very limited
3. I dont get what unit test has to be created to cover the deprecation/legacy group stuff requested by @OskarStark 

so in the end, the amount of work I have to put into this is not aligned with the benefit of the outcome. Thank you all for the guiding, sorry if I wasted your time with this.

stof commented:
@boesing writing a unit test that pass a Input implementation without `__toString` to the Application, and expecting a deprecation message.

derrabus commented:
@OskarStark Do you have time to take over?
